File: src/shared/ipc-contracts/device.ts
/**
 * Wire contract for the `read_connected_lunii` Tauri command. Mirror of
 * `src-tauri/src/ipc/dto/device.rs::ConnectedDeviceDto`.
 *
 * Shape: tagged enum on `kind` ∈ `"none" | "supported" | "unsupported"
 * | "ambiguous"`. Every payload field is camelCase; `firmwareCohort`
 * and `reason` use camelCase string enums (Rust serde
 * `rename_all = "camelCase"`). Cross-stack contract tests keep the wire
 * shape symmetric.
 */

export type SupportedFamilyDto = "lunii" | "flam";

export type FirmwareCohortDto = "origineV1" | "midGenV2" | "v3" | "flamGen1";

export interface SupportedOperationsDto {
  readLibrary: boolean;
  inspectStory: boolean;
  importStory: boolean;
  writeStory: boolean;
  /** Delete a story already on the device. Gated separately from
   *  `writeStory`: deletion removes opaque bytes and needs no
   *  pack-format ciphering, so Lunii V3 may delete even while it may
   *  not (yet) be written to. */
  deleteStory: boolean;
  /** Send a STUdio-format pack archive (`.zip`) to the device. Gated
   *  separately from `writeStory` (the round-trip of an imported pack):
   *  the archive-send owns its whole V3 pipeline, so Lunii V3 may
   *  receive archives while the round-trip stays closed. */
  sendArchive: boolean;
}

export type UnsupportedReasonDto =
  | "firmwareUnsupported"
  | "metadataUnsupported"
  | "metadataCorrupt"
  | "familyUnknown"
  | "operationNotAuthorized"
  | "multipleCandidates";

export type ConnectedDeviceDto =
  | { kind: "none" }
  | {
      kind: "supported";
      family: SupportedFamilyDto;
      firmwareCohort: FirmwareCohortDto;
      /** Present for families whose primary marker carries a version
       *  byte (Lunii). ABSENT — the key itself, never `null` — for
       *  families without one (FLAM): the Rust serializer omits it. */
      metadataFormatVersion?: number;
      deviceIdentifier: string;
      supportedOperations: SupportedOperationsDto;
    }
  | {
      kind: "unsupported";
      reason: UnsupportedReasonDto;
      firmwareHint: string | null;
    }
  | { kind: "ambiguous"; candidateCount: number };

/** Closed per-family contract: which cohorts are legal for the family
 *  and whether `metadataFormatVersion` must be PRESENT (an integer in
 *  0..127) or ABSENT (the key itself — JSON has no `undefined`, and
 *  `null` is refused). Replaces the former independent family/cohort
 *  sets: an illegal combination (`lunii`+`flamGen1`, `flam` carrying a
 *  version, …) is unrepresentable at the boundary. */
const FAMILY_CONTRACTS: Record<
  string,
  { cohorts: ReadonlySet<string>; metadataFormatVersion: "required" | "absent" }
> = {
  lunii: {
    cohorts: new Set(["origineV1", "midGenV2", "v3"]),
    metadataFormatVersion: "required",
  },
  flam: {
    cohorts: new Set(["flamGen1"]),
    metadataFormatVersion: "absent",
  },
};
const UNSUPPORTED_REASONS: ReadonlySet<string> = new Set([
  "firmwareUnsupported",
  "metadataUnsupported",
  "metadataCorrupt",
  "familyUnknown",
  "operationNotAuthorized",
  "multipleCandidates",
]);

function isSupportedOperationsDto(
  value: unknown,
): value is SupportedOperationsDto {
  if (typeof value !== "object" || value === null) return false;
  const c = value as Record<string, unknown>;
  return (
    typeof c.readLibrary === "boolean" &&
    typeof c.inspectStory === "boolean" &&
    typeof c.importStory === "boolean" &&
    typeof c.writeStory === "boolean" &&
    typeof c.deleteStory === "boolean" &&
    typeof c.sendArchive === "boolean"
  );
}

/** 32 lowercase hex chars — must mirror exactly what the Rust core
 *  emits via `compute_device_identifier`. A non-matching string is a
 *  protocol drift, not a transient quirk. */
const DEVICE_IDENTIFIER_PATTERN = /^[0-9a-f]{32}$/;

/** Closed set of permitted property names per `kind`. Used to refuse
 *  payloads that carry extra fields — a drifted Rust serializer
 *  must fail the guard loudly, not be quietly tolerated. */
const ALLOWED_KEYS: Record<string, ReadonlySet<string>> = {
  none: new Set(["kind"]),
  supported: new Set([
    "kind",
    "family",
    "firmwareCohort",
    "metadataFormatVersion",
    "deviceIdentifier",
    "supportedOperations",
  ]),
  unsupported: new Set(["kind", "reason", "firmwareHint"]),
  ambiguous: new Set(["kind", "candidateCount"]),
};

/** Prototype-safe own-property lookup. A plain indexation
 *  (`record[key]`) walks the prototype chain: a hostile discriminant
 *  such as `"constructor"` or `"__proto__"` would resolve to a truthy
 *  `Object.prototype` member and crash the guard with a `TypeError`
 *  instead of a boolean rejection. */
function ownEntry<T>(record: Record<string, T>, key: string): T | undefined {
  return Object.prototype.hasOwnProperty.call(record, key)
    ? record[key]
    : undefined;
}

function hasOnlyAllowedKeys(
  value: Record<string, unknown>,
  kind: string,
): boolean {
  const allowed = ownEntry(ALLOWED_KEYS, kind);
  if (!allowed) return false;
  for (const k of Object.keys(value)) {
    if (!allowed.has(k)) return false;
  }
  return true;
}

/**
 * Runtime guard for `ConnectedDeviceDto`. Rejects every drift: unknown
 * `kind`, missing fields, extra fields, wrong types, unrecognized enum
 * strings, and malformed `deviceIdentifier` payloads. The UI must
 * never render against an arbitrary object — a drift is a fail-loud
 * bug, not a silent fallback.
 */
export function isConnectedDeviceDto(
  value: unknown,
): value is ConnectedDeviceDto {
  if (typeof value !== "object" || value === null) return false;
  const c = value as Record<string, unknown>;
  if (typeof c.kind !== "string") return false;
  if (!hasOnlyAllowedKeys(c, c.kind)) return false;
  switch (c.kind) {
    case "none":
      return true;
    case "supported": {
      if (typeof c.family !== "string") return false;
      const contract = ownEntry(FAMILY_CONTRACTS, c.family);
      if (!contract) return false;
      if (
        typeof c.firmwareCohort !== "string" ||
        !contract.cohorts.has(c.firmwareCohort)
      )
        return false;
      if (contract.metadataFormatVersion === "required") {
        if (
          typeof c.metadataFormatVersion !== "number" ||
          !Number.isInteger(c.metadataFormatVersion) ||
          c.metadataFormatVersion < 0 ||
          c.metadataFormatVersion > 127
        )
          return false;
      } else {
        // ABSENT means the KEY is absent: a present key — even
        // `null`/`undefined`-valued — is a producer drift, refused.
        if ("metadataFormatVersion" in c) return false;
      }
      if (
        typeof c.deviceIdentifier !== "string" ||
        !DEVICE_IDENTIFIER_PATTERN.test(c.deviceIdentifier)
      )
        return false;
      if (!isSupportedOperationsDto(c.supportedOperations)) return false;
      return true;
    }
    case "unsupported":
      if (
        typeof c.reason !== "string" ||
        !UNSUPPORTED_REASONS.has(c.reason)
      )
        return false;
      if (c.firmwareHint !== null && typeof c.firmwareHint !== "string")
        return false;
      return true;
    case "ambiguous":
      return (
        typeof c.candidateCount === "number" &&
        Number.isInteger(c.candidateCount) &&
        c.candidateCount >= 2
      );
    default:
      return false;
  }
}

File: src-tauri/tests/contracts/device.rs
use rustory_lib::ipc::dto::{
    ConnectedDeviceDto, FirmwareCohortDto, SupportedFamilyDto, SupportedOperationsDto,
    UnsupportedReasonDto,
};

fn supported_dto() -> ConnectedDeviceDto {
    ConnectedDeviceDto::Supported {
        family: SupportedFamilyDto::Lunii,
        firmware_cohort: FirmwareCohortDto::OrigineV1,
        metadata_format_version: Some(3),
        device_identifier: "0123456789abcdef0123456789abcdef".into(),
        supported_operations: SupportedOperationsDto {
            read_library: true,
            inspect_story: true,
            import_story: true,
            write_story: false,
            delete_story: true,
            send_archive: false,
        },
    }
}

fn supported_flam_dto() -> ConnectedDeviceDto {
    ConnectedDeviceDto::Supported {
        family: SupportedFamilyDto::Flam,
        firmware_cohort: FirmwareCohortDto::FlamGen1,
        metadata_format_version: None,
        device_identifier: "fedcba9876543210fedcba9876543210".into(),
        // The FLAM Gen1 matrix line: read capabilities ✅✅✅, write ❌,
        // delete ❌, archive-send ❌ (unproven on real FLAM hardware).
        supported_operations: SupportedOperationsDto {
            read_library: true,
            inspect_story: true,
            import_story: true,
            write_story: false,
            delete_story: false,
            send_archive: false,
        },
    }
}

#[test]
fn connected_device_none_serializes_with_kind_none() {
    let v = serde_json::to_value(&ConnectedDeviceDto::None).expect("ser");
    assert_eq!(v, serde_json::json!({ "kind": "none" }));
}

#[test]
fn connected_device_supported_origine_v1_round_trip_wire_shape() {
    let v = serde_json::to_value(supported_dto()).expect("ser");
    assert_eq!(
        v,
        serde_json::json!({
            "kind": "supported",
            "family": "lunii",
            "firmwareCohort": "origineV1",
            "metadataFormatVersion": 3,
            "deviceIdentifier": "0123456789abcdef0123456789abcdef",
            "supportedOperations": {
                "readLibrary": true,
                "inspectStory": true,
                "importStory": true,
                "writeStory": false,
                "deleteStory": true,
                "sendArchive": false,
            },
        })
    );
}

#[test]
fn connected_device_supported_v3_serializes_with_import_story_false() {
    let dto = ConnectedDeviceDto::Supported {
        family: SupportedFamilyDto::Lunii,
        firmware_cohort: FirmwareCohortDto::V3,
        metadata_format_version: Some(7),
        device_identifier: "fedcba9876543210fedcba9876543210".into(),
        supported_operations: SupportedOperationsDto {
            read_library: true,
            inspect_story: true,
            import_story: false,
            write_story: false,
            delete_story: true,
            send_archive: true,
        },
    };
    let v = serde_json::to_value(&dto).expect("ser");
    assert_eq!(v["firmwareCohort"], "v3");
    assert_eq!(v["supportedOperations"]["importStory"], false);
    assert_eq!(v["supportedOperations"]["writeStory"], false);
    // V3 CAN delete even while import/write stay closed — deletion is
    // crypto-free (delist + remove opaque bytes).
    assert_eq!(v["supportedOperations"]["deleteStory"], true);
    // And V3 CAN receive a pack archive: the DEDICATED send capability,
    // open while the round-trip `writeStory` stays closed.
    assert_eq!(v["supportedOperations"]["sendArchive"], true);
}

#[test]
fn connected_device_unsupported_metadata_serializes_with_typed_reason() {
    let dto = ConnectedDeviceDto::Unsupported {
        reason: UnsupportedReasonDto::MetadataUnsupported,
        firmware_hint: Some("metadata_v99".into()),
    };
    let v = serde_json::to_value(&dto).expect("ser");
    assert_eq!(v["kind"], "unsupported");
    assert_eq!(v["reason"], "metadataUnsupported");
    assert_eq!(v["firmwareHint"], "metadata_v99");
}

#[test]
fn connected_device_unsupported_serializes_null_firmware_hint_when_absent() {
    let dto = ConnectedDeviceDto::Unsupported {
        reason: UnsupportedReasonDto::MetadataCorrupt,
        firmware_hint: None,
    };
    let v = serde_json::to_value(&dto).expect("ser");
    assert_eq!(v["reason"], "metadataCorrupt");
    assert!(v["firmwareHint"].is_null());
}

#[test]
fn connected_device_ambiguous_with_2_candidates() {
    let dto = ConnectedDeviceDto::Ambiguous { candidate_count: 2 };
    let v = serde_json::to_value(&dto).expect("ser");
    assert_eq!(v["kind"], "ambiguous");
    assert_eq!(v["candidateCount"], 2);
}

#[test]
fn supported_operations_dto_uses_camel_case_only() {
    let v = serde_json::to_value(supported_dto()).expect("ser");
    let ops = &v["supportedOperations"];
    for camel in [
        "readLibrary",
        "inspectStory",
        "importStory",
        "writeStory",
        "deleteStory",
        "sendArchive",
    ] {
        assert!(ops.get(camel).is_some(), "missing camelCase field: {camel}");
    }
    for snake in [
        "read_library",
        "inspect_story",
        "import_story",
        "write_story",
        "delete_story",
        "send_archive",
    ] {
        assert!(
            ops.get(snake).is_none(),
            "snake_case must not leak: {snake}"
        );
    }
}

#[test]
fn firmware_cohort_serializes_as_camel_case_string_for_each_variant() {
    for (variant, expected) in [
        (FirmwareCohortDto::OrigineV1, "origineV1"),
        (FirmwareCohortDto::MidGenV2, "midGenV2"),
        (FirmwareCohortDto::V3, "v3"),
    ] {
        let v = serde_json::to_value(&variant).expect("ser");
        assert_eq!(v, serde_json::Value::String(expected.into()));
    }
}

#[test]
fn unsupported_reason_serializes_as_camel_case_string_for_each_variant() {
    for (variant, expected) in [
        (
            UnsupportedReasonDto::FirmwareUnsupported,
            "firmwareUnsupported",
        ),
        (
            UnsupportedReasonDto::MetadataUnsupported,
            "metadataUnsupported",
        ),
        (UnsupportedReasonDto::MetadataCorrupt, "metadataCorrupt"),
        (UnsupportedReasonDto::FamilyUnknown, "familyUnknown"),
        (
            UnsupportedReasonDto::OperationNotAuthorized,
            "operationNotAuthorized",
        ),
        (
            UnsupportedReasonDto::MultipleCandidates,
            "multipleCandidates",
        ),
    ] {
        let v = serde_json::to_value(&variant).expect("ser");
        assert_eq!(v, serde_json::Value::String(expected.into()));
    }
}

#[test]
fn none_variant_does_not_emit_extra_fields() {
    let v = serde_json::to_value(&ConnectedDeviceDto::None).expect("ser");
    let obj = v.as_object().expect("object");
    // Exactly one key: "kind".
    assert_eq!(obj.len(), 1);
    assert!(obj.contains_key("kind"));
}

#[test]
fn device_identifier_field_is_string_not_object() {
    let v = serde_json::to_value(supported_dto()).expect("ser");
    assert!(v["deviceIdentifier"].is_string());
}

#[test]
fn connected_device_supported_does_not_emit_snake_case_aliases() {
    let v = serde_json::to_value(supported_dto()).expect("ser");
    for snake in [
        "firmware_cohort",
        "metadata_format_version",
        "device_identifier",
        "supported_operations",
    ] {
        assert!(v.get(snake).is_none(), "snake_case must not leak: {snake}");
    }
}

/// Byte-for-byte INVARIANCE of the Lunii supported wire: the exact JSON
/// string frozen as a literal. `deleteStory` then `sendArchive` were each
/// appended as deliberate wire extensions (device-delete, then archive-send
/// capability) AFTER `writeStory`, keeping every prior field's position and
/// value; the `Option<u8>` migration of `metadataFormatVersion` still must
/// not deform the wire — key present, plain integer, same order.
#[test]
fn connected_device_supported_lunii_wire_string_is_byte_for_byte_unchanged() {
    let s = serde_json::to_string(&supported_dto()).expect("ser");
    assert_eq!(
        s,
        "{\"kind\":\"supported\",\"family\":\"lunii\",\"firmwareCohort\":\"origineV1\",\
         \"metadataFormatVersion\":3,\
         \"deviceIdentifier\":\"0123456789abcdef0123456789abcdef\",\
         \"supportedOperations\":{\"readLibrary\":true,\"inspectStory\":true,\
         \"importStory\":true,\"writeStory\":false,\"deleteStory\":true,\
         \"sendArchive\":false}}"
    );
}

/// Twin of the invariance test for the FLAM wire (re-scoped with the
/// activated read capabilities, never deleted): `family`/`firmwareCohort`
/// discriminate, `readLibrary`/`inspectStory`/`importStory` are `true`,
/// `writeStory` stays `false`, and the `metadataFormatVersion` key is
/// ABSENT from the string (never `null`).
#[test]
fn connected_device_supported_flam_wire_string_omits_version_key() {
    let s = serde_json::to_string(&supported_flam_dto()).expect("ser");
    assert_eq!(
        s,
        "{\"kind\":\"supported\",\"family\":\"flam\",\"firmwareCohort\":\"flamGen1\",\
         \"deviceIdentifier\":\"fedcba9876543210fedcba9876543210\",\
         \"supportedOperations\":{\"readLibrary\":true,\"inspectStory\":true,\
         \"importStory\":true,\"writeStory\":false,\"deleteStory\":false,\
         \"sendArchive\":false}}"
    );
    assert!(!s.contains("metadataFormatVersion"));
    assert!(!s.contains("null"));
}

#[test]
fn firmware_cohort_flam_gen1_serializes_as_camel_case_string() {
    let v = serde_json::to_value(FirmwareCohortDto::FlamGen1).expect("ser");
    assert_eq!(v, serde_json::Value::String("flamGen1".into()));
}

#[test]
fn supported_family_flam_serializes_as_camel_case_string() {
    let v = serde_json::to_value(SupportedFamilyDto::Flam).expect("ser");
    assert_eq!(v, serde_json::Value::String("flam".into()));
}
